src/file.cpp

```diff
@@ -321,8 +321,28 @@ void File::terminated()
 // failuresup to 16 times, except I/O errors (6 times)
 bool File::failed(error e)
 {
-    return (e != API_EKEY && e != API_EBLOCKED && e != API_ENOENT && e != API_EINTERNAL && transfer->failcount < 16) &&
-            !((e == API_EREAD || e == API_EWRITE) && transfer->failcount > 6);
+    if (e == API_EKEY)
+    {
+        if (!transfer->currentmetamacvalid)
+        {
+            // several integrity check errors uploading chunks
+            return transfer->failcount < 1;
+        }
+
+        if (transfer->prevmetamacvalid && transfer->prevmetamac == transfer->currentmetamac)
```

Comment:
Maybe refactor `prevmetamacvalid` and `currentmetamacvalid` to something that doesn't mean _valid_, like `hasprevmetamac` or similar.